src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/HiviUap.java

```diff
@@ -0,0 +1,105 @@
+package com.wikia.webdriver.pageobjectsfactory.componentobject.ad;
+
+import com.wikia.webdriver.common.core.WikiaWebDriver;
+import com.wikia.webdriver.common.core.elemnt.Wait;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.interactions.Actions;
+
+import java.util.concurrent.Callable;
+
+public class HiviUap {
+
+  private final Wait wait;
+  private final String slot;
+
+  private static final String SLOT_SELECTOR = "#%s ";
+  private static final String FULL_SCREEN_BUTTON_SELECTOR = ".toggle-fullscreen-button";
+  private static final String REPLAY_SELECTOR = ".replay-overlay";
+  private static final String VIDEO_CLICK_AREA_SELECTOR = ".toggle-ui-overlay";
+  private static final String VIDEO_THUMBNAIL_SELECTOR = ".videoThumbnail";
+  private static final String PAUSE_SELECTOR = ".play-pause-button";
+  private static final String SOUND_SELECTOR = ".volume-button";
+  private static final String VIDEO_PLAYER = ".video-player";
+
+  private final WikiaWebDriver driver;
+
+  public HiviUap(WikiaWebDriver driver, String slot) {
+    this.wait = new Wait(driver);
+    this.driver = driver;
+    this.slot = slot;
+  }
+
+  public void waitForVideoStart() {
+    usingAdFrame(() -> wait.forElementNotVisible(By.cssSelector(VIDEO_THUMBNAIL_SELECTOR)));
+  }
+
+  public void waitForVideoEnd() {
+    usingAdFrame(() -> wait.forElementVisible(By.cssSelector(String.format(REPLAY_SELECTOR, slot)), 60));
+  }
+
+  private void usingAdFrame(Runnable f) {
+    WebElement adIframe = wait.forElementPresent(By.cssSelector(
+        String.format(SLOT_SELECTOR, slot) + " iframe[id^=\"google_ads\"]"
+    ));
+    driver.switchTo().frame(adIframe);
+    try {
+      f.run();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-s4/coding_rules#rule_key=squid%3AS00112)